File: app/src/main/java/ir/arapp/arappmain/view/utils/SingleScrollDirectionEnforcer.java
package ir.arapp.arappmain.view.utils;

import android.view.MotionEvent;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import static java.lang.Math.abs;

public class SingleScrollDirectionEnforcer extends RecyclerView.OnScrollListener implements RecyclerView.OnItemTouchListener {

    private int scrollState = RecyclerView.SCROLL_STATE_IDLE;
    private int scrollPointerId = -1;
    private int initialTouchX = 0;
    private int initialTouchY = 0;
    private int dx = 0;
    private int dy = 0;


    @Override
    public void onScrollStateChanged(@NonNull @NotNull RecyclerView recyclerView, int newState) {
        super.onScrollStateChanged(recyclerView, newState);
        int oldState = scrollState;
        scrollState = newState;
        if (oldState == RecyclerView.SCROLL_STATE_IDLE && newState == RecyclerView.SCROLL_STATE_DRAGGING) {
            RecyclerView.LayoutManager layoutManager = recyclerView.getLayoutManager();
            if (layoutManager != null) {
                boolean canScrollHorizontally = layoutManager.canScrollHorizontally();
                boolean canScrollVertically = layoutManager.canScrollVertically();
                if (canScrollHorizontally != canScrollVertically) {
                    if ((canScrollHorizontally && abs(dy) > abs(dx))
                            || (canScrollVertically && abs(dx) > abs(dy))) {
                        recyclerView.stopScroll();
                    }
                }
            }
        }
    }

    @Override
    public boolean onInterceptTouchEvent(@NonNull @NotNull RecyclerView rv, @NonNull @NotNull MotionEvent e) {
        if (e.getActionMasked() == MotionEvent.ACTION_DOWN) {
            scrollPointerId = e.getPointerId(0);
            initialTouchX = (int) (e.getX() + 0.5f);
            initialTouchY = (int) (e.getY() + 0.5f);
        } else if (e.getActionMasked() == MotionEvent.ACTION_POINTER_DOWN) {
            int actionIndex = e.getActionIndex();
            scrollPointerId = e.getPointerId(actionIndex);
            initialTouchX = (int) (e.getX(actionIndex) + 0.5f);
            initialTouchY = (int) (e.getY(actionIndex) + 0.5f);
        } else if (e.getActionMasked() == MotionEvent.ACTION_UP) {
            int index = e.findPointerIndex(scrollPointerId);
            if (index >= 0 && scrollState != RecyclerView.SCROLL_STATE_DRAGGING) {
                int x = (int) (e.getX(index) + 0.5f);
                int y = (int) (e.getY(index) + 0.5f);
                dx = x - initialTouchX;
                dy = y - initialTouchY;
            }
        }
        return false;
    }

    @Override
    public void onTouchEvent(@NonNull @NotNull RecyclerView rv, @NonNull @NotNull MotionEvent e) {

    }

    @Override
    public void onRequestDisallowInterceptTouchEvent(boolean disallowIntercept) {

    }
}

File: app/src/main/java/ir/arapp/arappmain/view/utils/SingleScrollDirectionEnforcer.kt
package ir.arapp.arappmain.view.utils

import android.view.MotionEvent
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.OnItemTouchListener

class SingleScrollDirectionEnforcer : RecyclerView.OnScrollListener(), OnItemTouchListener {
    private var scrollState = RecyclerView.SCROLL_STATE_IDLE
    private var scrollPointerId = -1
    private var initialTouchX = 0
    private var initialTouchY = 0
    private var dx = 0
    private var dy = 0
    override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
        super.onScrollStateChanged(recyclerView, newState)
        val oldState = scrollState
        scrollState = newState
        if (oldState == RecyclerView.SCROLL_STATE_IDLE && newState == RecyclerView.SCROLL_STATE_DRAGGING) {
            val layoutManager = recyclerView.layoutManager
            if (layoutManager != null) {
                val canScrollHorizontally = layoutManager.canScrollHorizontally()
                val canScrollVertically = layoutManager.canScrollVertically()
                if (canScrollHorizontally != canScrollVertically) {
                    if (canScrollHorizontally && Math.abs(dy) > Math.abs(dx)
                        || canScrollVertically && Math.abs(dx) > Math.abs(dy)
                    ) {
                        recyclerView.stopScroll()
                    }
                }
            }
        }
    }

    override fun onInterceptTouchEvent(rv: RecyclerView, e: MotionEvent): Boolean {
        if (e.actionMasked == MotionEvent.ACTION_DOWN) {
            scrollPointerId = e.getPointerId(0)
            initialTouchX = (e.x + 5f).toInt()
            initialTouchY = (e.y + 5f).toInt()
        } else if (e.actionMasked == MotionEvent.ACTION_POINTER_DOWN) {
            val actionIndex = e.actionIndex
            scrollPointerId = e.getPointerId(actionIndex)
            initialTouchX = (e.getX(actionIndex) + 5f).toInt()
            initialTouchY = (e.getY(actionIndex) + 5f).toInt()
        } else if (e.actionMasked == MotionEvent.ACTION_UP) {
            val index = e.findPointerIndex(scrollPointerId)
            if (index >= 0 && scrollState != RecyclerView.SCROLL_STATE_DRAGGING) {
                val x = (e.getX(index) + 5f).toInt()
                val y = (e.getY(index) + 5f).toInt()
                dx = x - initialTouchX
                dy = y - initialTouchY
            }
        }
        return false
    }

    override fun onTouchEvent(rv: RecyclerView, e: MotionEvent) {}
    override fun onRequestDisallowInterceptTouchEvent(disallowIntercept: Boolean) {}
}